larpandora/LArPandoraEventBuilding/LArPandoraShower/Algs/LArPandoraShowerAlg.cxx

```diff
@@ -360,6 +360,83 @@ double shower::LArPandoraShowerAlg::SpacePointPerpendicular(art::Ptr<recob::Spac
   return pos.Mag();
 }
 
+  //Function to calculate the RMS at segments of the shower and calculate the gradient of this. If negative then the direction is pointing the opposite way to the correct one
+  double shower::LArPandoraShowerAlg::RMSShowerGradient(std::vector<art::Ptr<recob::SpacePoint> >& sps, const TVector3& ShowerCentre, const TVector3& Direction, const unsigned int nSegments) const {
+
+    if (nSegments == 0)
+      throw cet::exception("LArPandoraShowerAlg") << "Unable to calculate RMS Shower Gradient with 0 segments" << std::endl;
+
+    if (sps.size() < 3)
+      return 0;
+
+    //Order the spacepoints
+    this->OrderShowerSpacePoints(sps,ShowerCentre,Direction);
+
+    //Get the length of the shower.
+    const double minProj = this->SpacePointProjection(sps[0],ShowerCentre,Direction);
+    const double maxProj = this->SpacePointProjection(sps[sps.size()-1],ShowerCentre,Direction);
+
+    const double length = (maxProj-minProj);
+    const double segmentsize = length / nSegments;
+
+    if (segmentsize < std::numeric_limits<double>::epsilon())
+      return 0;
+
+    std::map<int, std::vector<float> > len_segment_map;
+
+    //Split the the spacepoints into segments.
+    for(auto const& sp: sps){
+
+      //Get the the projected length
+      const double len = this->SpacePointProjection(sp,ShowerCentre,Direction);
+
+      //Get the length to the projection
+      const double  len_perp = this->SpacePointPerpendicular(sp,ShowerCentre,Direction,len);
+
+      int sg_len = std::round(len/segmentsize);
+      len_segment_map[sg_len].push_back(len_perp);
+    }
+
+    int counter = 0;
+    float sumx  = 0;
+    float sumy  = 0;
+    float sumx2 = 0;
+    float sumxy = 0;
+
+    //Get the rms of the segments and caclulate the gradient.
+    for(auto const& segment: len_segment_map){
+      if (segment.second.size()<2) continue;
+      float RMS = this->CalculateRMS(segment.second);
+
+      if (RMS < 0)
+        continue;
+
+      //Calculate the gradient using regression
+      sumx  += segment.first;
+      sumy  += RMS;
+      sumx2 += segment.first * segment.first;
+      sumxy += RMS * segment.first;
+      ++counter;
+    }
+
+    const float denom(counter*sumx2 - sumx*sumx);
```

Comment:
you initialise here but assign elsewhere.  I think initialisation is generally better but its probably even better to be consistent everywhere.